### PR TITLE
KOGITO-1236: Adding Preview API

### PR DESCRIPTION
**Pull request by jesuino**
PR to allow editors to provide a preview SVG of the running diagram.

In this PR we create a new message in the envelope API and expose a method to send and consume this message. Get preview changes according to the channel:

* DESKTOP: Preview will be called for recently opened files and users can get a preview using the application menu;
* VSCODE: a new command allow users to get a snapshot;
* Online Editor: A new button in kebab allow users to download the SVG of the displayed diagram.

Part of an ensemble, merge with:

https://github.com/kiegroup/kie-wb-common/pull/3192
https://github.com/kiegroup/appformer/pull/906

### CODE REVIEW

**Comment by jesuino** [comment]
@tiagobento I did rebase this PR and tested from end to end and everything seems to continue working after the rebase. Would you please review?

Thanks!

File: packages/chrome-extension/src/app/components/common/KogitoEditorIframe.tsx [hunk 1]
```diff
@@ -106,6 +106,9 @@ const RefForwardingKogitoEditorIframe: React.RefForwardingComponent<IsolatedEdit
         },
         receive_resourceListRequest(pattern: string) {
           resourceContentService.list(pattern).then(list => self.respond_resourceList(list));
+        },
+        receive_preview(previewSVG: string) {
+          console.debug("received preview: " + previewSVG ) ;
```
Comment (tiagobento) [inline packages/chrome-extension/src/app/components/common/KogitoEditorIframe.tsx]: Could we change that to `previewSvg`? It's a matter of personal preference, but I've personally been writing acronyms  in camel/Pascal case on this code base and maybe we should stick to this "convention"? 😝

File: packages/microeditor-envelope-protocol/src/EnvelopeBusMessageType.ts [hunk 2]
```diff
@@ -34,6 +34,9 @@ export enum EnvelopeBusMessageType {
   REQUEST_RESOURCE_LIST,
   RETURN_RESOURCE_LIST,
 
+  REQUEST_PREVIEW,
+  RESPOND_PREVIEW,
```
Comment (tiagobento) [inline packages/microeditor-envelope-protocol/src/EnvelopeBusMessageType.ts]: Convention so far has been REQUEST and RETURN.

File: packages/kie-bc-editors/src/__tests__/DummyEditor.tsx [hunk 3]
```diff
@@ -41,6 +41,10 @@ export class DummyEditor extends AppFormer.Editor {
   public setContent(content: string) {
     return this.ref!.setContent(content);
   }
+
+  public getPreview(): Promise<string> {
+    return Promise.resolve("");
+  }
```
Comment (tiagobento) [inline packages/kie-bc-editors/src/__tests__/DummyEditor.tsx]: Does empty string mean 'No preview'?

File: packages/microeditor-envelope/src/EditorEnvelopeController.tsx [hunk 4]
```diff
@@ -86,6 +86,15 @@ export class EditorEnvelopeController {
...
         this.stateControl.redo();
+      },
+      receive_previewRequest: () => {
+        const editor = this.getEditor();
+        if (editor) {
+          editor
+            .getPreview()
+            .then(preview => self.respond_previewRequest(preview))
+            .catch(error => console.log(`Error retrieving preview: ${error}`));
+        }
```
Comment (tiagobento) [inline packages/microeditor-envelope/src/EditorEnvelopeController.tsx]: Here we could easily use Optional chaining. If you want to change the others methods too, feel free. I've written the other ones before TypeScript 3.7 , so there was no Optional chaining available at the time.

File: packages/online-editor/src/editor/Editor.tsx [hunk 5]
```diff
@@ -23,10 +23,12 @@ import { EditorContent, ResourceContent, ResourcesList, ResourceContentRequest }
 interface Props {
   fullscreen: boolean;
   onContentResponse: (content: EditorContent) => void;
+  onPreviewResponse: (preview: string) => void;
```
Comment (tiagobento) [inline packages/online-editor/src/editor/Editor.tsx]: Rename to `previewSvg`?

File: packages/online-editor/src/editor/Editor.tsx [hunk 6]
```diff
@@ -66,6 +68,10 @@ const RefForwardingEditor: React.RefForwardingComponent<EditorRef, Props> = (pro
       receive_resourceListRequest(globPattern: string) {
         console.debug(`Resource List Request`);
         self.respond_resourceList(new ResourcesList(globPattern, []));
+      },
+      receive_preview(preview: string) {
+        console.debug("received preview");
+        props.onPreviewResponse(preview);
       }
```
Comment (tiagobento) [inline packages/online-editor/src/editor/Editor.tsx]: Rename to `previewSvg`?

File: packages/microeditor-envelope-protocol/src/EnvelopeBusOuterMessageHandler.ts [hunk 7]
```diff
@@ -110,6 +111,10 @@ export class EnvelopeBusOuterMessageHandler {
     this.busApi.postMessage({ type: EnvelopeBusMessageType.RETURN_RESOURCE_LIST, data: resourcesList });
   }
 
+  public request_preview() {
+    this.busApi.postMessage({ type: EnvelopeBusMessageType.REQUEST_PREVIEW, data: undefined });
+  }
```
Comment (tiagobento) [inline packages/microeditor-envelope-protocol/src/EnvelopeBusOuterMessageHandler.ts]: This should be called `request_previewResponse`. Go with the flow! 😝

File: packages/microeditor-envelope-protocol/src/EnvelopeBusOuterMessageHandler.ts [hunk 8]
```diff
@@ -35,6 +35,7 @@ export interface EnvelopeBusOuterMessageHandlerImpl {
   receive_dirtyIndicatorChange(isDirty: boolean): void;
   receive_resourceContentRequest(resourceContentService: ResourceContentRequest): void;
   receive_resourceListRequest(globPattern: string): void;
+  receive_preview(previewSVG: string) : void;
```
Comment (tiagobento) [inline packages/microeditor-envelope-protocol/src/EnvelopeBusOuterMessageHandler.ts]: This one should be `receive_previewRequest`.

File: packages/online-editor/src/editor/EditorPage.tsx [hunk 9]
```diff
@@ -48,6 +49,7 @@ export function EditorPage(props: Props) {
   const history = useHistory();
   const editorRef = useRef<EditorRef>(null);
   const downloadRef = useRef<HTMLAnchorElement>(null);
+  const previewRef = useRef<HTMLAnchorElement>(null);
```
Comment (tiagobento) [inline packages/online-editor/src/editor/EditorPage.tsx]: `downloadPreviewRef`?

File: packages/vscode-extension-pack-kogito-kie-editors/package.json [hunk 10]
```diff
@@ -20,6 +20,12 @@
   "publisher": "kiegroup",
   "main": "./dist/extension/extension.js",
   "contributes": {
+    "commands": [
+      {
+        "command": "extension.kogito.getPreview",
```
Comment (tiagobento) [inline packages/vscode-extension-pack-kogito-kie-editors/package.json]: `extension.kogito.getPreviewSvg`?

File: packages/vscode-extension-pack-kogito-kie-editors/package.json [hunk 11]
```diff
@@ -20,6 +20,12 @@
...
+        "title": "Get Editor Preview"
```
Comment (tiagobento) [inline packages/vscode-extension-pack-kogito-kie-editors/package.json]: `Get Editor Preview SVG`?

**Review by tiagobento — CHANGES_REQUESTED** [review]
Looks very good! Some minor comments inline. Good job @jesuino!

**Review by jomarko — COMMENTED** [review]
I tried to run the VS code, but after running "Get Editor Preview" command no file was generated.

**Comment by jesuino** [comment]
Hello @tiagobento 

I made the changes you requested, would you please check it again? I tested from end to end and it is working fine after all the refactor.

**Comment by jesuino** [comment]
> I tried to run the VS code, but after running "Get Editor Preview" command no file was generated.

@jomarko The upstream version was incremented to 7.35 SNAPSHOT. You must:

* build appformer on this branch;
* build kie-wb-common on this branch, or at least, build the kogito runtimes.
* Increment the version in `kie-bc-editors-unpacked/pom.xml` to 7.35 SNAPSHOT 
* Build kogito-tooling

Then it should work on online editor and on VSCode.

Thanks and let me know if you face any issue.

**Review by jomarko — COMMENTED** [review]
There is no notification, once svg is created, is that intentional?

**Comment by jomarko** [comment]
@jesuino colors are not included in the generated svg. Is that intentional?

**Comment by jesuino** [comment]
Hello @jomarko 

>> There is no notification, once svg is created, is that intentional?

Yes, I didn't add a notification because it was too annoying. @tiagobento @jomarko let me know if you think that a notification saying that the file was generated is useful and I will add it. Thanks! :)

**Comment by jesuino** [comment]
@jomarko 

>> @jesuino colors are not included in the generated svg. Is that intentional?

Could you please clarify what you mean by colors? For me the generated SVG seems okay.

Thanks!

**Comment by jomarko** [comment]
@jesuino if the original diagram is colored, the svg do not contain colors.
Screenshot from 2020-03-03 14-02-11

**Review by jomarko — APPROVED** [review]
OK reported and approving this one

https://issues.redhat.com/browse/RHDM-1280

**Review by tiagobento — APPROVED** [review]
Thanks @jesuino!! Looks ready now :)

**Review by ederign — APPROVED** [review]

**Pull request merged**